src/blog/better-class-features.md

```diff
@@ -0,0 +1,332 @@
+
+## Optimizing private methods
+
+### The implementation of private methods
+
+In [the specification](https://tc39.es/ecma262/#sec-privatemethodoraccessoradd), the private methods are described as if they are installed on the instances but not on the class. In order to save memory, however, V8's implementation stores the private methods along with a private brand symbol in a context associated with the class. When the constructor is invoked, V8 only stores a reference to that context in the instance, with the private brand symbol as the key.
+
+![Evaluation and instantiation of classes with private methods](/_img/better-class-features/class-evaluation-and-instantiation.svg)
+
+When the private methods are accessed, V8 walks the context chain starting from the execution context to find the class context, reads a statically known slot from the found context to get the private brand symbol for the class, then checks if the instance has a property keyed by this brand symbol to see if the instance is created from this class. If the brand check passes, V8 loads the private method from another known slot in the same context and completes the access.
+
+![Access of private methods](/_img/better-class-features/access-private-methods.svg)
+
+Take this snippet as an example:
+
+```js
+class A {
+  #a() {}
+}
+```
+
+V8 used to generate the following bytecode for the constructor of `A`:
+
+```cpp
+// Load the private brand symbol for class A from the context
+// and store it into r1.
+LdaImmutableCurrentContextSlot [3]
+Star r1
+
+// Load the target into r0.
+Mov <this>, r0
+// Load the current context into r2.
+Mov <context>, r2
+// Call the runtime %AddPrivateBrand() function to store the context in
+// the instance with the private brand as key.
+CallRuntime [AddPrivateBrand], r0-r2
+```
+
+Since there was also a call to the runtime function `%AddPrivateBrand()`, the overhead made the constructor much slower than constructors of classes with only public methods.
+
+### Optimizing initialization of private brands
+
+To speed up the installation of the private brands, in most cases we just reuse the `DefineKeyedOwnProperty` machinery added for the optimization of private fields:
+
+```cpp
+// Load the private brand symbol for class A from the context
+// and store it into r1
+LdaImmutableCurrentContextSlot [3]
+Star0
+
+// Use the DefineKeyedOwnProperty bytecode to store the
+// context in the instance with the private brand as key
+Ldar <context>
+DefineKeyedOwnProperty <this>, r0, [0]
+```
+
+![Performance of initialization of instances with private methods](/_img/better-class-features/private-methods-performance.svg)
```

Comment:
Ditto on these graphs to label the OS + arch + general performance profile info.